RevDiff: compare any two files

gerhardol opened the pull request:
Fixes #8048 

## Proposed changes

 * If two items in the RevDiff list are selected, add menu item to compare them
 * If one item is selected, menu item to remember the file
 * If one item is selected and one item remembered, menu item to compare them

This allows to compare refactored files, between commits, currently you have to save temporary files and compare them
The primary use is refactoring, when files are renamed and changed without Git detecting the change or when the change is done over several commits. It also allows seeing the difference for images (like in p4merge) when files are renamed.

Viewing files requires that the file exist in the commit, the first item must not be work tree.
To simplify usage, the compared file is automatically selecting the first revision if the second revision does cannot be used for the diff.

See https://github.com/gitextensions/gitextensions/pull/8193#issuecomment-643833240 for some use cases.

The first and second revisions are not named, just First and Second as the revisions are explained in the top of the menu.
 
Refactoring:
 * DiffTool should always detect rename and copy
Hard to track that "-M -C" is always used, the parameters are passed around in some situations, in the Git command for  others

 * RevDiff: Allow difftool menu for untracked
Working in some difftools, comparing to NUL

 * [New] Use _Second_ instead of _Selected_ in the Difftool submenu (should have been changed when the naming for the groups were changed).

 * [New] Keep _Second_ above _First_ in the difftool submenu (already this order in Reset submenu)

Possible enhancement:
Add _Remember file_ and _Diff with <remembered file>_ in FileTree

## Screenshots 

### Before

![image](https://user-images.githubusercontent.com/6248932/83955309-d48c4000-a851-11ea-9d44-03996193a8b1.png)

### After

![image](https://user-images.githubusercontent.com/6248932/85205841-22a74600-b31e-11ea-9525-52c25899f3a6.png)

![image](https://user-images.githubusercontent.com/6248932/85903036-af15a500-b805-11ea-8766-e34452af31ec.png)

![image](https://user-images.githubusercontent.com/6248932/85903110-d8363580-b805-11ea-91d5-31c70a42c68e.png)

(Change from Selected to _Second_, same as in Diff submenu)
![image](https://user-images.githubusercontent.com/6248932/85205907-aa8d5000-b31e-11ea-99f5-3ce60d371b79.png)


## Test methodology <!-- How did you ensure quality? -->

Added tests

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented:
👍 Works well. I'll review the code after the rebase.

I'd like to add a separator above the new menu items.
`Save selected file to later` should be rephrased to e.g. `Save selected file for later comparison`.
The filename should be quoted in `Compare selected file to "file"`.
Perhaps this menu item should be suppressed if the identical file and revision is selected. Git does not open the difftool in this case.

I think it would be useful to have these context menu items in the file tree, too. E.g. if an unchanged file was cloned to a new one.

Slightly off-topic for this PR: I still quarrel over the new position of the submenu `Open with difftool`.
It even does not have a fixed position because
`Cherry pick changes` is hidden for multi-selections. It should rather be disabled if it is too difficult to implement it.

gerhardol commented:
> Works well. I'll review the code later (edit: has already been rebased).

I will rebase at least when there are conflicts, more often if I want to sync to other PRs.

> I'd like to add a separator above the new menu items.
Save selected file to later should be rephrased to e.g. Save selected file for later comparison.
The filename should be quoted in Compare selected file to "file".
Perhaps this menu item should be suppressed if the identical file and revision is selected. Git does not open the difftool in this case.

Agree. I have to leave something for you too

> I think it would be useful to have these context menu items in the file tree, too. E.g. if an unchanged file was cloned to a new one.

Can this be separator or do we need a separate resolver that is common?

I probably want to have an alternative for _Save parent to selected for later comparison_ (nice and short? Should have the revision too). With a divider it may be OK to see these separate 

> Slightly off-topic for this PR: I still quarrel over the new position of the submenu Open with difftool.
It even does not have a fixed position because
Cherry pick changes is hidden for multi-selections. It should rather be disabled if it is too difficult to implement it.

(I do not like the position either, but found it better than an implementation that was inconsistent.)
Implementation right now is to get the diff from the diff viewer. It would be better to get the diff (patch) separately anyway and apply that (whitespaces etc). 

A bigger problem is to implement the fish-needs-bicycle tests for the controller.

RussKie commented on GitUI/CommandsDialogs/FileStatusListContextMenuController.cs:
```diff
@@ -103,5 +115,53 @@ public bool ShouldDisplayMenuSelectedParentToLocal(ContextMenuDiffToolInfo selec
+        }
+
+        /// <inheritdoc/>>
+        public string GetGitCommit([CanBeNull] GitModule module, [CanBeNull] FileStatusItem item, bool isFirst)
```
When `GitModule` is passed around it makes me very uneasy. This adds unnecessary coupling.
How about this?
```suggestion
        public string GetGitCommit([CanBeNull] FileStatusItem item, [CanBeNull] Func<string, ObjectId, ObjectId> getFileBlobHash, bool isFirst)
```

Call-sites should be:
```cs
var first = _revisionDiffContextMenuController.GetGitCommit(diffFiles[1], Module.GetFileBlobHash, true);
```

And this should simplify testing as well.

gerhardol replied:
will change

RussKie commented:
What does "Save selected file to later" mean?

gerhardol commented:
> What does "Save selected file to later" mean?

Save the file+commit so it can be compared later. Really similar to comparing the two commits, then compare, but it may be easier to find the files/commits to compare.
Similar to the explorer context menus for difftools like BeyondCompare, DiffMerge and TortoiseGit.

RussKie commented:
> Similar to the explorer context menus for difftools like BeyondCompare, DiffMerge and TortoiseGit.

Do they have the same wording?
May be it should be _for_ - _"Save selected file for later"_? 

...and even then it looks/sounds awkward.

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -742,6 +783,17 @@ private void openWithDifftoolToolStripMenuItem_DropDownOpening(object sender, Ev
+            compareTwoSelectedDifftoolToolStripMenuItem.Enabled = diffFiles.Count == 2
+                                                                  && _revisionDiffContextMenuController.ShouldEnableFirstSpecialCompare(diffFiles?[1])
+                                                                  && _revisionDiffContextMenuController.ShouldEnableSecondSpecialCompare(diffFiles?[0]);
+            compareToExistingDifftoolToolStripMenuItem.Visible = diffFiles.Count == 1 && _savedCompareFileItem != null;
+            compareToExistingDifftoolToolStripMenuItem.Enabled = diffFiles.Count == 1 && _revisionDiffContextMenuController.ShouldEnableSecondSpecialCompare(diffFiles?[0]);
+            compareToExistingDifftoolToolStripMenuItem.Text = _savedCompareFileItem != null ? string.Format(_compareSelectedFile.Text, _savedCompareFileItem.Item.Name) : string.Empty;
+            saveForLaterDifftoolToolStripMenuItem.Visible = diffFiles.Count == 1;
+            saveForLaterDifftoolToolStripMenuItem.Enabled = diffFiles.Count == 1 && _revisionDiffContextMenuController.ShouldEnableFirstSpecialCompare(diffFiles?[0]);
```
I like the idea of being able to compare arbitrary files across revisions, but I'm finding the visual presentation of the feature confusing. I had to look at the code to understand the screenshots showed...

Few Qs that immediately popped up:
1. how do I know which files from which revisions I marked?
2. how do I unmark selected/all files?
3. how do I change the order of files?

mstv replied:
> how do I know which files from which revisions I marked?

The file name is shown in the menu item `Compare selected file to "file"`.
Its revision is something for your short-term memory :)

>    how do I change the order of files?

Click ![grafik](https://user-images.githubusercontent.com/36601201/84541147-17896000-acf7-11ea-92ed-75b4ad66963f.png)
Or replace the selection using `Save selected file for later comparison` and then chose `Compare selected file to "file"` after selecting the first one again.

>    how do I unmark selected/all files?

It's all about comparing exactly two files.
No need to unmark, just start over with `Save selected file for later comparison`.

gerhardol replied:
I want to have the previous selectable too
```Diff the selected files``` (when two selected)
```Diff with "file"``` (one selected, revision not included in text)
```Save b/12345678 for later``` (one selected, filename not included)
```Save a/12345678 for later```

Screenshots for other apps. No consistent description. The proposals are considerably longer...
 * No selection
 * Select the same file again 
 * Another selected
 * Two selected
 * more than two selected
![image](https://user-images.githubusercontent.com/6248932/84543230-3b4ea500-acfb-11ea-8213-377eb45216d1.png)

DiffMerge can save several files, as well as clear them
![image](https://user-images.githubusercontent.com/6248932/84547880-b4062f00-ad04-11ea-9512-052998629749.png)

Tortoise supports both git-diff and file system from the context menu
![image](https://user-images.githubusercontent.com/6248932/84547662-4fe36b00-ad04-11ea-8555-a9bda1ccfb40.png)
 
![image](https://user-images.githubusercontent.com/6248932/84543399-96809780-acfb-11ea-8f88-a13782a9c13c.png)

RussKie replied:
Thank you, it's starting to make sense.
Now to pick and agree on the wording :) I think if I saw BC and TG menus I'd conceivably  guess their intent.
Would "Remember file for compare" work? And then "Compare with <the remembered file>"?

gerhardol replied:
(I missed this comment, waiting with a merge)
 The menu and Git uses diff rather than compare, so I prefer diff.
"Compare selected..." should be "Diff with <saved file>" (dropping "selected file".

But "Remember Second for later" or "Remember Second" (similar to First) is possible. The shorter, the better is my preference (I constantly fail though.)

@mstv ?

gerhardol commented:
I have pushed changes to private branches in my repo only (will test some more)
9d8989256d63a730b8f43f31a5b19652bbd22c16 contains the differences, will squash that, see 18e14f654472655d3611992c4e9d7c1b28bc14ba
Description and screenshots will be updated

This cleans up the the handling of assuming  first rev when second rev cannot be used, including addig "Save first revision" to the context menu.
(Prepared for mstv's suggestion to add to the context menu there, not done though, awaiting review first.)

Use case 1:
File is renamed but Git do not detect the change
5dae3022679a3b65dd6bffd0eb5054f220a67933
Use P4Merge or some tool that display images
Diff IconAbout.co and information.ico

Use case 2:
Compare similar files
dc89d30157eb966c3f2d8a1bf337489655697af7

Use case 3:
(My primary use, but also the most complicated)
Refactoring splitting a file (there are better examples that I do not find now)
19129fbc2a7d90c41d1be87b2f88081560e2aa93
Select FormBrowse.cs, First, then RevisionDiff.cs, Diff with.

gerhardol commented on GitCommands/Git/GitModule.cs:
```diff
@@ -3654,6 +3648,31 @@ public string OpenWithDifftool(string filename, string oldFileName = "", string
+        /// <param name="firstGitCommit">commitish</param>
+        /// <param name="secondGitCommit">commitish</param>
+        /// <returns>empty string</returns>
+        public string OpenFilesWithDifftool(string firstGitCommit, string secondGitCommit)
```
It follows the pattern for OpenWithDifftool()
See the TODO there
As an empty string is always returned anyway, few callers try to use the data
This should be handled the same

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -742,6 +803,39 @@ private void openWithDifftoolToolStripMenuItem_DropDownOpening(object sender, Ev
+            diffTwoSelectedDifftoolToolStripMenuItem.Visible = diffFiles.Count == 2;
+            diffTwoSelectedDifftoolToolStripMenuItem.Enabled =
+                diffFiles.Count == 2
+                && (_saveForLaterContextMenuController.ShouldEnableFirstItemDiff(diffFiles[firstIndex], false)
+                    || _saveForLaterContextMenuController.ShouldEnableFirstItemDiff(diffFiles[firstIndex], true))
```
Do you want ShouldEnableFirstItemFirstRevisionDiff() and ShouldEnableFirstItemSecondRevisionDiff()?
I believe this is harder to understand

mstv replied:
I mean
```
bool ShouldEnableFirstItemDiff(diffFile)
    => ShouldEnableFirstItemDiff(diffFile, flag: false)
    || ShouldEnableFirstItemDiff(diffFile, flag: true);

bool ShouldEnableFirstItemDiff(diffFile, bool flag)
{
    ...
}
```

gerhardol replied:
ok

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -67,6 +69,11 @@ private void FileViewer_BottomScrollReached(object sender, EventArgs e)
             DiffText.ScrollToTop();
         }
 
+        public void UICommands_PostRepositoryChanged(object sender, GitUIEventArgs e)
+        {
+            _saveForLaterContextMenuController.SavedDiffFileItem = null;
```
In this case the item is used in the GUI only, so no real difference.
As this affects FormBrowse, I suggest skipping a change

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -630,6 +637,60 @@ private void openWithDifftoolToolStripMenuItem_Click(object sender, EventArgs e)
+            var firstIndex = DiffFiles.SelectedItem == diffFiles[0] ? 1 : 0;
+
+            // Fallback to first revision if second revision cannot be used
+            var isFirstItemSecondRev = _saveForLaterContextMenuController.ShouldEnableFirstItemDiff(diffFiles[firstIndex], true);
```
Related to your ShouldEnableFirstItemSecondRevisionDiff() suggestion below, which one do you suggest?

mstv replied:
just add the name of the flag, please: (Re# is too slow and not available at GH)

gerhardol replied:
ok (added also where the flag is called isSecondRevision, except for test cases)

gerhardol commented on GitCommands/Git/GitModule.cs:
```diff
@@ -3633,10 +3627,9 @@ public IEnumerable<string> GetPreviousCommitMessages(int count, string revision
         }
 
-        public string OpenWithDifftoolDirDiff(string firstRevision = GitRevision.IndexGuid, string secondRevision = GitRevision.WorkTreeGuid, string extraDiffArguments = null)
+        public string OpenWithDifftoolDirDiff(string firstRevision, string secondRevision)
```
above

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -51,9 +52,10 @@ public RevisionDiffControl()
             DiffText.TopScrollReached += FileViewer_TopScrollReached;
             DiffText.BottomScrollReached += FileViewer_BottomScrollReached;
-        }
+         }
```
ok

gerhardol commented:
> Found an edge case: No matter whether the first or second of a renamed file is saved for later, the new name appears in the menu:

I had ignored that, but since you are commenting on it I will change that
Sidenote: Save for later is intentionally not disabling the menuitem if already selected, intentional

> And: Should it be `First -> Second` instead of `First -> Selected`?

Yes

> Should `Second b/...` rather be the top item in order to match the order in the RevisionGrid and in the `Reset file(s) to` menu?

It should be the same as in Reset at least. That Selected was before First was natural, but First before Second may not be.
But if the natural selection should be first, the order should be Send, First also in Reset.

mstv commented:
> Sidenote: Save for later is intentionally not disabling the menuitem if already selected

👍 else I would have complained 😉

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -630,6 +637,60 @@ private void openWithDifftoolToolStripMenuItem_Click(object sender, EventArgs e)
+        {
+            // For first item, the second revision is explicitly saved
+            var first = _saveForLaterContextMenuController.GetGitCommit(Module.GetFileBlobHash,
+                _saveForLaterContextMenuController.SavedDiffFileItem, true);
```
another unlabeled `true`

gerhardol commented:
> > Should `Second b/...` rather be the top item in order to match the order in the RevisionGrid and in the `Reset file(s) to` menu?
 
> It should be the same as in Reset at least. That Selected was before First was natural, but First before Second may not be.
> But if the natural selection should be first, the order should be Send, First also in Reset.

Added a commit to align diff submenu with reset submenu, updated description and screenshot, also removed WIP

RussKie commented:
I had a brief looked at the code, it looks good, I have no objections.

gerhardol commented:
@msftbot merge in 24 hours

RussKie commented:
For the ease of navigation, continuation of the discussion in https://github.com/gitextensions/gitextensions/pull/8193#discussion_r439253418

Something like this?

* "Diff the selected files" (when two selected)
* "Diff with \<remembered file>" (visible, if there is a 'remembered' file)
* separator
* "Remember b/12345678 for diff` (one selected, filename not included)
* "Remember a/12345678 for diff`
* "Forget \<remembered file>" (visible, if there is a 'remembered' file)

gerhardol commented:
Adapted the proposal.

> separator 

(No need to have a separator further down "Diff the selected" is only shown with two selected, the other only with one

> "Diff the selected files" (when two selected)
> "Diff with <remembered file>" (when one file selected) (visible, if there is a 'remembered' file)
> "Remember Second for diff` (one selected, filename not included)
> "Remember First for diff`

Second and First is explained in the top of the menu, no need to do it again

> "Forget <remembered file>" (visible, if there is a 'remembered' file)

No need for this one, forgotten if swiching repos.
I consider this as junk in the menu actions, more confusing than helping

...And the references to "saveForLater" in the code should be changed to remember file...

RussKie commented:
👍 
So then, just "Save" -> "Remember"

Thank you

mstv commented:
I agree.
Though I noticed today: Compared to the menu items above the separator, `Compare selected file to "remembered file"` could / should be simplified to `Second -> "remembered file"`. If you find this too reduced, then I suggest to avoid the term "Compare to" and rather use "Diff with", which seems to be preferred by git and GE.

> Is "first" and "second" placement material?

It has been made consistent with the other submenus, and matches the order in the RevisionGrid.
Sorry this comes really late: Unfortunately, it is opposite to the display of patches. Which should take precedence: DiffView or RevisionGrid? Which is nearer?

gerhardol commented:
> It has been made consistent with the other submenus, and matches the order in the RevisionGrid.
Sorry this comes really late: Unfortunately, it is opposite to the display of patches. Which should take precedence: DiffView or RevisionGrid? Which is nearer?

At this point, I say follow RevisionGrid "one selected" order

Will squash latest commit in a day or so (no changes other than naming in the latest commit)
(The first four commits should remain)
If no further comments, I will set a merge a day after that

#8239 will follow, then likely #8254 before #7825 can be resumed

RussKie commented:
Works for me. Thanks

On Sat, Jun 27, 2020, 7:48 AM Gerhard Olsson <notifications@github.com>
wrote:

> It has been made consistent with the other submenus, and matches the order
> in the RevisionGrid.
> Sorry this comes really late: Unfortunately, it is opposite to the display
> of patches. Which should take precedence: DiffView or RevisionGrid? Which
> is nearer?
>
> At this point, I say follow RevisionGrid "one selected" order
>
> Will squash latest commit in a day or so (no changes other than naming in
> the latest commit)
> (The first four commits should remain)
> If no further comments, I will set a merge a day after that
>
> #8239 <https://github.com/gitextensions/gitextensions/pull/8239> will
> follow, then likely #8254
> <https://github.com/gitextensions/gitextensions/pull/8254> before #7825
> <https://github.com/gitextensions/gitextensions/pull/7825> can be resumed
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8193#issuecomment-650418279>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXSFXQFGIEVUOTUGQTLRYUJSHANCNFSM4NWNJEYQ>
> .
>
